frontend/components/BookForm.tsx

```diff
@@ -0,0 +1,120 @@
+import { AuthorsData } from "@/types/author";
+import { Book } from "@/types/book";
+import { GET_AUTHORS } from "@/utils/graphql";
+import { useQuery } from "@apollo/client";
+import { Form, Modal, Input, Select, Spin } from "antd";
+import moment from "moment";
+import React, { useState } from "react";
+
+const { Option } = Select;
+
+const BookForm = ({
+  visible,
+  setVisible,
+  onCreate,
+  data = null,
+  setDataToNull,
+}: {
+  visible: boolean;
+  setVisible: React.Dispatch<React.SetStateAction<boolean>>;
+  onCreate: (values: any) => void;
+  setDataToNull: () => void;
+  data?: null | Book;
+}) => {
+  const [form] = Form.useForm();
+  const {
+    data: authorsData,
+    loading: authorsLoading,
+    error: authorsError,
+  } = useQuery<AuthorsData>(GET_AUTHORS);
+
+  React.useEffect(() => {
+    if (data) {
+      form.setFieldsValue({
+        ...data,
+        published_date: moment(parseInt(data.published_date)).format(
+          "YYYY-MM-DD"
+        ),
+        author_id: data?.author?.id || "",
+      });
+    } else {
+      form.resetFields();
+    }
+  }, [data, form]);
+
+  const handleCreate = () => {
+    form
+      .validateFields()
+      .then((values) => {
+        form.resetFields();
+        onCreate(values);
+      })
+      .catch((info) => {
+        console.log("Validate Failed:", info);
+      });
+  };
+
+  return (
+    <Modal
+      visible={visible}
+      title={data ? "Edit Book" : "Create a new book"}
```

Comment:
Generally is a good idea to reuse same component for 2 different actions?  How would we do tradeoff here?

Reply:
Usually should not reuse the same component.